include/wil/networking.h

```diff
@@ -0,0 +1,1487 @@
+//! These headers have many intra-header dependencies, creating difficulties when needing access to various functions and types
+//! This specific sequence should compile correctly to give access to all available functions and types
+#include <winsock2.h>
+#include <ws2def.h>
+#include <ws2ipdef.h>
+#include <mswsock.h>
+#include <mstcpip.h>
+#include <ws2tcpip.h>
+#include <windns.h>
+#include <iphlpapi.h>
+
+// the wil header for RAII types
+#include "resource.h"
+
+namespace wil
+{
+//! Functions and classes that support networking operations and structures
+namespace networking
+{
+    class socket_address;
+    //! A type that calls WSACleanup on destruction (or reset()).
+    //! WSAStartup must be called for the lifetime of all Winsock APIs (synchronous and asynchronous)
+    //! WSACleanup will unload the full Winsock catalog - all the libraries - with the final reference
+    //! which can lead to crashes if socket APIs are still being used after the final WSACleanup is called
+    using unique_wsacleanup_call = ::wil::unique_call<decltype(&::WSACleanup), ::WSACleanup>;
+
+    //! Calls WSAStartup; returns an RAII object that reverts, the RAII object will resolve to bool 'false' if failed
+    WI_NODISCARD inline unique_wsacleanup_call WSAStartup_nothrow() WI_NOEXCEPT
+    {
+        WSADATA unused_data{};
+        const auto error = ::WSAStartup(WINSOCK_VERSION, &unused_data);
+        LOG_IF_WIN32_ERROR(error);
+
+        unique_wsacleanup_call return_cleanup{};
+        if (error != 0)
+        {
+            // internally set m_call to false
+            // so the caller can check the return object against its operator bool
+            // to determine if WSAStartup succeeded
+            return_cleanup.release();
+        }
+        return return_cleanup;
+    }
+
+    //! Calls WSAStartup and fail-fasts on error; returns an RAII object that reverts
+    WI_NODISCARD inline unique_wsacleanup_call WSAStartup_failfast() WI_NOEXCEPT
+    {
+        WSADATA unused_data{};
+        FAIL_FAST_IF_WIN32_ERROR(::WSAStartup(WINSOCK_VERSION, &unused_data));
+        return {};
+    }
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+    //! Calls WSAStartup and throws on error; returns an RAII object that reverts
+    WI_NODISCARD inline unique_wsacleanup_call WSAStartup()
+    {
+        WSADATA unused_data{};
+        THROW_IF_WIN32_ERROR(::WSAStartup(WINSOCK_VERSION, &unused_data));
+        return {};
+    }
+#endif
+
+    struct WINSOCK_EXTENSION_FUNCTION_TABLE
+    {
+        LPFN_ACCEPTEX AcceptEx{nullptr};
+        LPFN_CONNECTEX ConnectEx{nullptr};
+        LPFN_DISCONNECTEX DisconnectEx{nullptr};
+        LPFN_GETACCEPTEXSOCKADDRS GetAcceptExSockaddrs{nullptr};
+        LPFN_TRANSMITFILE TransmitFile{nullptr};
+        LPFN_TRANSMITPACKETS TransmitPackets{nullptr};
+        LPFN_WSARECVMSG WSARecvMsg{nullptr};
+        LPFN_WSASENDMSG WSASendMsg{nullptr};
+    };
+
+    struct winsock_extension_function_table

```

Comment:
Why is this needed? Why not rely on the usual build time import linkage and IAT?